JT/12_data_structures/non-lineat_types_exercises/non_linear_types.adoc

```diff
@@ -0,0 +1,288 @@
+== Tree
+:icons: font
+image::../resources/tree.png[width=700,align="center"]
+
+A *tree* is a collection of elements (nodes) connected with edges that reflect parent-child relationships.
+Each node (except the root—the topmost node) has a single parent.
+Nodes without children are known as leaves.
+
+*Types of Trees*::
+There are many types of trees that we can use to organize data differently within a hierarchical structure.
+The tree we use depends on the problem we are trying to solve.
+
+|===
+|Tree type |Description |Example
+
+|N-ary trees
+|In N-ary tree, a node can have child nodes from 0-N
+a|image::../resources/n_ary_tree.png[]
+
+|Balanced trees
+|A balanced tree is a tree with almost all leaf nodes at the same level
+a|image::../resources/balanced_tree.png[]
+
+|Binary trees
+|All interior nodes should have two children, and all leaves must have the same depth.
+Look at this diagram to understand how a perfect binary tree looks.
+a|image::../resources/perfect_binary_tree.png[]
+
+|Red-Black Trees
+a|
+
+* The root is always black
+* Two red nodes cannot be adjacent (i.e. a red parent cannot have a red child)
+* A path from the root to a leaf should contain the same number of black nodes
+* A null node is black
+a|image::../resources/red_black_tree.png[]
+|===
+
+.*How to Create the Tree?*
+You can use any class that implements the Queue interface, like LinkedList or PriorityQueue.
+
+----
+public class TreeNode {
```

Comment:
I would suggest to add java formatting with `[source,java]` for every java code that we have in material. Like this:
```
[source,java]
----
public class TreeNode {

  private int value;
  private List<TreeNode> childNodes;

  public TreeNode(int value) {
    this.value = value;
    this.childNodes = new ArrayList<>();
  }
TreeNode root = new TreeNode(1);
----
```

Reply:
Done